crates/llama-cpp-bindings/src/engine.cc

```diff
@@ -256,15 +280,16 @@ class TextInferenceEngineImpl : public TextInferenceEngine {
 
         int32_t i_batch = request.i_batch - i;
         float* logits = llama_get_logits_ith(ctx, i_batch);
-        compute_softmax_inplace(logits, n_vocab, request.temperature);
-        const llama_token next_token = weighted_random(logits, n_vocab, request.seed);
+        compute_softmax_inplace(logits, n_vocab, request.temperature());
+        const llama_token next_token = weighted_random(ctx_.get(), &request, logits, n_vocab);
         request.n_past += request.tokens.size();
 
         request.tokens.clear();
         request.tokens.push_back(next_token);
 
         const auto token_str = llama_token_to_piece(ctx, next_token);
         request.generated_text += token_str;
+        request.accept_candidate(token_str);
```

Comment:
@boxbeam when an token got selected, we accept it at end.

Reply:
This is where I think we should have a loop to get new tokens until it finds one that works better. I think giving it a cap on how many attempts it can take (3-10 might be a good default) would be a good idea.